### ISSUE TITLE
Should there be a makeArray primitive, and for what?

### ISSUE THREAD

**Issue by Bastacyclop** [rise-lang/rise#27]
Currently we have a `makeArray` primitive that works very much like an array literal, except that it works with elements that are expressions, instead of plain data like literals do. Currently we only generate code for `makeArray` if indices are constant (we know which element to pick at codegen time).

1. ~~Why can our normal array literals not contain arbitrary expressions?~~
2. **Should we generate code for non-constant indices, and how?**
3. **Why do we need array literals when we can use array constructors like makeArray?**

Personally, I would like to solve both of these questions by removing the `makeArray` primitive. Instead of plain data, normal array literals would take elements of expression kind.

Additonnaly, a `selectArray` -- or differently named high-level construct -- would also allow to create arrays as a combination of `generate` and `select`. This would generate C code with if statements or ternaries to pick the right element. If indices are constant, we should simplify the branching away.

@rise-lang/core

**Comment by bastian-koepcke** [comment]
1. A literal is always a value. The name would not be correct anymore if it was build out of expressions (e.g., `5` value, `2 + 3` expression).
2. Would the combination of `generate` and `select` be able to generate code for non-constant indices, because `generate` already uses non-constant indices?

**Comment by Bastacyclop** [comment]
>     2. Would the combination of `generate` and `select` be able to generate code for non-constant indices, because `generate` already uses non-constant indices?

Yes, I already descibed that above. In more details for a 2-element array this would look like:
```
generate (i => select (i = 0) first second)
```
